segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/filesystem/IdempotentStorageBase.java

```diff
@@ -0,0 +1,103 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.impl.filesystem;
+
+import com.google.common.base.Preconditions;
+import io.pravega.common.Exceptions;
+import io.pravega.segmentstore.storage.Storage;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Base class for all the storage implementations that support idempotent write at an offset.
+ */
+public abstract class IdempotentStorageBase implements Storage {
```

Comment:
1. What is so Idempotent about this? I remember asking to extract some common methods into a helper (final) class. No need to subclass it.
2. This is the wrong location for it. This is used outside of this package, which means `extendeds3` package depends on this package.
3. Looks like the only common method relate to futures. Can you figure out if you can reuse something from `FutureHelpers`, or otherwise put it/them in there and get rid of this class?